### PR TITLE
 fix: warn if dependencies are missing when generating

### PR DESCRIPTION
**Pull request by hampuslavin**
## Description
Warns if there are packages in the `<app>_server` module that are not installed, which fixes [#2384](https://github.com/serverpod/serverpod/issues/2384).

Also captures exceptions at the top level and surfaces sad-paths that were failing silently.

## Changes
- Adds a warning if packages are not installed
- Prints out exceptions

## Pre-launch Checklist
- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

### CODE REVIEW

File: tools/serverpod_cli/lib/src/config/config.dart [hunk 1]
```diff
@@ -152,41 +153,33 @@ class GeneratorConfig {
...
-      log.error(
-        'Failed to load pubspec.yaml. Are you running serverpod from your '
-        'projects root directory?',
-      );
-
       throw const ServerpodProjectNotFoundException(
-        'Failed to load valid pubspec.yaml',
+        'Failed to load pubspec.yaml. Are you running serverpod from your '
+        'projects server root directory?',
       );
```
Comment (hampuslavin) [inline tools/serverpod_cli/lib/src/config/config.dart:169]: Confusing with two error messages, where one is never surfaced. Thought it was clearer to just have a single error message.

File: tools/serverpod_cli/lib/src/config/config.dart [hunk 2]
```diff
@@ -283,6 +275,25 @@ class GeneratorConfig {
...
+
+    if (packageConfig == null) {
+      throw const ServerpodProjectNotFoundException(
+        'Failed to read your server\'s package configuration. Have you run '
+        '`dart pub get` in your server directory?',
+      );
+    }
+
+    var allPackagesAreInstalled = pubspec.dependencies.keys
+        .every((dependencyName) => packageConfig[dependencyName] != null);
```
Comment (hampuslavin) [inline tools/serverpod_cli/lib/src/config/config.dart:290]: Is there a case for checking `dev_dependencies` as well?

Comment (Isakdl) [reply tools/serverpod_cli/lib/src/config/config.dart:290]: As of today, no since we only care about modules here and they must be normal dependencies.

**Review by Isakdl — CHANGES_REQUESTED** [review]
Looking good! Many nice modifications here! :) 

Small little suggestion on code structure.

File: tools/serverpod_cli/lib/src/config/config.dart [hunk 3]
```diff
@@ -217,15 +210,14 @@ class GeneratorConfig {
...
+      throw const ServerpodProjectNotFoundException(
         'Failed to load client pubspec.yaml. If you are using a none default '
         'path it has to be specified in the config/generator.yaml file!',
       );
-      throw const ServerpodProjectNotFoundException(
-        'Failed to load client pubspec.yaml',
-      );
     }
 
+    await _warnIfDependenciesAreNotInstalled(serverRootDir, pubspec);
```
[the remark this answers is not in the document]
Comment (hampuslavin) [reply tools/serverpod_cli/lib/src/config/config.dart:219]: Good point! Will fix 👍

[the remark this answers is not in the document]
Comment (hampuslavin) [reply tools/serverpod_cli/lib/src/config/config.dart]: 5bc91a7

**Review by Isakdl — APPROVED** [review]
LGTM!

**Pull request merged**